Convert Schneider Electric shutters to quirks v2, expose entities

uvNikita opened the pull request:
## Proposed change
- Convert to quirks v2
- Define missing attributes using spec doc
- Add support for extra models (NHPB/SHUTTER/1 and PUCK/SHUTTER/1)
- Expose extra entities to home assistant

Fixes #1685
Supersedes #3939

## Additional information
#1705 
[ZB Spec - Micro Module Shutter Blinds - 110422.pdf](https://github.com/user-attachments/files/20774274/ZB.Spec.-.Micro.Module.Shutter.Blinds.-.110422.pdf)


## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [ ] The changes are tested and work correctly
- [ ] `pre-commit` checks pass / the code has been formatted using Black
- [ ] Tests have been added to verify that the new code works

uvNikita committed:
Convert Schneider Electric shutters to quirks v2, expose entities

Add support for NHPB/SHUTTER/1 and PUCK/SHUTTER/1

uvNikita commented:
Unfortunately I didn't see https://github.com/zigpy/zha-device-handlers/pull/3939 until creating the PR, but this one should include everything #3939 has + quirks v2, extra models and exposed entities.

Mookunicorn commented:
Do you have any idea if it's possible to add a choice for the curtain type? It's currently "Tilt_blind_tilt_and_lift" when it should be an awning?

I saw in the [documentation that you posted](https://github.com/user-attachments/files/20774274/ZB.Spec.-.Micro.Module.Shutter.Blinds.-.110422.pdf) that only the following values ​​can be chosen:
0x00 = Rollershade (LIFT ONLY)
0x08 = Tilt Blind-Lift and Tilt (LIFT AND TILT)

In my case, the Rollershade would be suitable for my situation, but when I try to change it via ZHA (on the Cluster "SEWindowCovering (Endpoint id: 5, Id: 0x0102, Type: in)" and the attribute "window_covering_type (id: 0x0000)" to 0, nothing happens.

uvNikita commented:
@Mookunicorn I don't have this shutter modules, so couldn't try it myself yet. But I'm planning to order them the following mouths.

Can you try to set the value to `WindowCoveringType.Rollershade` in "Manage Zigbee Device"? If it works, I can add this option as an entity too.

uvNikita commented:
@Mookunicorn 

Actually, it looks like this attribute is readonly, but in the spec doc it says the following:

> The WindowCoveringType attribute identifies the type of window covering being controlled by this endpoint.
If you set TiltOpenCloseAndStepTime attribute to 0, WindowCoveringType is set automaticaly to 0 (lift only), if TiltOpenCloseAndStepTime attribute is non zero, value of
WindowCoveringType attribute is automaticaly set 8. Meaning of values is following

Have you tried setting `TiltOpenCloseAndStepTime` to 0?

Mookunicorn commented:
> @Mookunicorn I don't have this shutter modules, so couldn't try it myself yet. But I'm planning to order them the following mouths.
> 
> Can you try to set the value to `WindowCoveringType.Rollershade` in "Manage Zigbee Device"? If it works, I can add this option as an entity too.

Ok it works using "WindowCoveringType.Rollershade" and I feel slightly stupid right now ☠️

Mookunicorn commented:
> @Mookunicorn
> 
> Actually, it looks like this attribute is readonly, but in the spec doc it says the following:
> 
> > The WindowCoveringType attribute identifies the type of window covering being controlled by this endpoint.
> > If you set TiltOpenCloseAndStepTime attribute to 0, WindowCoveringType is set automaticaly to 0 (lift only), if TiltOpenCloseAndStepTime attribute is non zero, value of
> > WindowCoveringType attribute is automaticaly set 8. Meaning of values is following
> 
> Have you tried setting `TiltOpenCloseAndStepTime` to 0?

After some tests, I noticed that changing the value of `TiltOpenCloseAndStepTime` to anything other than 0 does not produce any changes. However, when I read the attribute via "Manage Zigbee Device", it appears as **Tilt_blind_tilt_and_lift**.

The same principle applies to Rollershade.

uvNikita commented:
So just to check if I understand it correctly.

- Writing the value `WindowCoveringType.Rollershade` to WindowCoveringType doesn't work, only reading it?
- Changing the value of `TiltOpenCloseAndStepTime` works, but seemingly doesn't change the type of the blinds until you reread `WindowCoveringType`?

Mookunicorn commented:
That's right, I can try in the evening to change the value of `TiltOpenCloseAndStepTime` and see how long it updates (if there is a delay).

TheJulianJES commented on tests/test_schneiderelectric.py:
```diff
@@ -1,145 +1,14 @@
-    device = zigpy_device_from_quirk(
-        zhaquirks.schneiderelectric.shutters.OneGangShutter1
-    )
-    window_covering_cluster = device.endpoints[5].window_covering
-
-    p = mock.patch.object(window_covering_cluster, "request", mock.AsyncMock())
-    with p as request_mock:
-        request_mock.return_value = (foundation.Status.SUCCESS, "done")
-
-        await window_covering_cluster.up_open()
-
-        assert request_mock.call_count == 1
-        assert request_mock.call_args[0][1] == (
-            WindowCovering.ServerCommandDefs.up_open.id
-        )
-
-
-async def test_1gang_shutter_1_lift_percentage_updates(zigpy_device_from_quirk):
-    """Asserts that updates to the ``current_position_lift_percentage`` attribute.
-
-    (e.g., by the device) invert the reported percentage value.
-    """
-
-    device = zigpy_device_from_quirk(
-        zhaquirks.schneiderelectric.shutters.OneGangShutter1
-    )
-    window_covering_cluster = device.endpoints[5].window_covering
-    cluster_listener = ClusterListener(window_covering_cluster)
-
-    window_covering_cluster.update_attribute(
-        WindowCovering.AttributeDefs.current_position_lift_percentage.id,
-        77,
-    )
-
-    assert len(cluster_listener.attribute_updates) == 1
-    assert cluster_listener.attribute_updates[0] == (
-        WindowCovering.AttributeDefs.current_position_lift_percentage.id,
-        23,  # 100 - 77
-    )
-    assert len(cluster_listener.cluster_commands) == 0
```
Ideally, we should keep the tests (except for the signature test). You should be able to use `zigpy_device_from_v2_quirk` for the tests. Might need to search the repo for some examples.

uvNikita replied:
Good point, done.

uvNikita commented:
- Added tests back
- Added `window_covering_type` read-only enum entity with `reporting_config` to hopefully help with automatic update of this attribute when `TiltOpenCloseAndStepTime` is updated

Mookunicorn commented:
> * Added `window_covering_type` read-only enum entity with `reporting_config` to hopefully help with automatic update of this attribute when `TiltOpenCloseAndStepTime` is updated

It exposes a new entity but it ends up being duplicated. `window_covering_type` was already read without quirks. And unfortunately neither entity updates when `TiltOpenCloseAndStepTime` changes.

<details>
  <summary>Screen</summary>

![image](https://github.com/user-attachments/assets/31e2ae81-63ab-4fa3-a7f1-6244325d0dfa)

</details>

uvNikita commented:
> It exposes a new entity but it ends up being duplicated. `window_covering_type` was already read without quirks. And unfortunately neither entity updates when `TiltOpenCloseAndStepTime` changes.

I was not aware of that, thank you for testing. Removed the duplicate entry again now.

Then I'm not sure what else we can do here. @TheJulianJES let me know if you have any ideas how to make sure `window_covering_type` is reread on `TiltOpenCloseAndStepTime` change. Otherwise I think it should be ready to be merged.

TheJulianJES commented on zhaquirks/schneiderelectric/shutters.py:
```diff
@@ -1,117 +1,131 @@
+        endpoint_id=21,
+        cluster_id=SESwitchConfiguration.cluster_id,
+        attribute_name=SESwitchConfiguration.AttributeDefs.se_switch_indication.name,
+        translation_key="switch_indication",
+        fallback_name="Switch indication",
```
Can we move `translation_key` and then `fallback_name` to the bottom (so as the last argument per method)?
That matches what we do for all other v2 quirks.

uvNikita replied:
Done

TheJulianJES reviewed (approved):
Looks good IMO. Thanks!

TheJulianJES commented:
Maybe open an issue for the `window_covering_type` stuff and tag me there and explain why `window_covering_type` should change there.

Orange-GT3 commented:
@uvNikita - sorry to reawaken the discussion here but I have just installed two Wiser Micro Blind Controllers that control two awnings therefore the tilt functionality is not needed.

I changed the value of `Tilt open close and step time` to zero and that updated the value of `window_covering_type` to `WindowCoveringType.Rollershade`. Should this automatically disable the tilt related entities and controls because I still see those controls on a `Tile` or `Mushroom Cover` card when opening the ZHA `More info` dialogue.

<img width="679" height="733" alt="image" src="https://github.com/user-attachments/assets/225b3d58-1aac-4861-a6cb-4cb7659a6ba9" /> 
<img width="681" height="721" alt="image" src="https://github.com/user-attachments/assets/37e51340-ea8f-4c84-a07e-c7c94d1ad743" />

The other oddity is the percentage readings that I am getting from the devices. When the awnings are closed, both cards show `Open - 100%` and when they are open (out since it is an awning), they show `Closed`. The two cards differ in the way the slider is displayed too:

<img width="519" height="553" alt="image" src="https://github.com/user-attachments/assets/f60c0c8b-8f26-4f5b-9fd8-71705b535567" />

In the above image, the Terrace is actually open (out) and the Garden is closed.

The slider on the `Tile Card` seems to reflect the true physical state, although the wording is wrong and the `Mushroom Card` has both the slider and the wording wrong. When using the buttons, on either card, I have to click the opposite to what would be expected, i.e. `Close cover` to open it and vice versa.

I apologise if all this is nothing to do with the quirk but I'm really confused at the moment and unsure how to safely set positions via automations.

Mookunicorn commented:
@Orange-GT3 Have you tried calibrating the awning as indicated in [the documentation](https://www.productinfo.schneider-electric.com/wiser_home/odace-connected-blind-control-switch_wiser_home_device-user-guide/English/Odace%20-%20Connected%20Blind%20Control%20Switch_Wiser_Home_Device%20user%20guide_0000825416.xml/$/Semi-AutomaticCalibration_Odace_WHCPT_0001147033)? That's what worked for me when it was displaying 255%.

Once done, try configuring the device via ZHA.

Orange-GT3 commented:
> @Orange-GT3 Have you tried calibrating the awning as indicated in [the documentation](https://www.productinfo.schneider-electric.com/wiser_home/odace-connected-blind-control-switch_wiser_home_device-user-guide/English/Odace%20-%20Connected%20Blind%20Control%20Switch_Wiser_Home_Device%20user%20guide_0000825416.xml/$/Semi-AutomaticCalibration_Odace_WHCPT_0001147033)? That's what worked for me when it was displaying 255%.
> 
> Once done, try configuring the device via ZHA.

Thanks for your reply @Mookunicorn, but that is a different device to what I have. Mine is the Micro module that sits in the wall behind a 3 position toggle switch: [Wiser Micro Module Blind](https://www.productinfo.schneider-electric.com/clipsal/wiser-micro-module-blind_device-user-guide_clp/English/Wiser%20Micro%20Module%20Blind_Clipsal_Device%20user%20guide_0001118618.xml). There is no calibration method described.

The 255% does seem to have disappeared now and I either see `Closed`, when it is open, and `Open - 100%` when it is closed.

Edit: I guess the closed state for a vertical blind on a window, i.e. providing shade, is the same as the awning being extended (open and providing shade).

uvNikita commented:
@Orange-GT3 I also now have this device and no "tilt" functionality on my blinds.

If I remember correctly what I had to do is to set `Tilt open close and step time` to zero (like you did), then window_covering_type changes, but, as you said, the native shade entity still had tilt controls.
After that, I had to force zha to update the entities by either re-reading SEWindowCovering -> window_covering_mode, and/or reconfigure the device in ZHA, and/or restart home assistant.
Eventually the tilt controls should be gone from the entity in home assistant.

After that, you need to calibrate the blinds so that zha will show correct percentage values.
Lift your blinds all the way up, measure how much time it takes to go all the way down, and put the value in the corresponding setting. Repeat the same for the opposite direction (I had to actually set different values for different directions to get the correct percentage readings).  Note that it always goes all the way up/down on the first run after calibration and after power loss.

If you set correct timings, you should be able to set the blinds position pretty precisely, at least in my experience.

This is how controls should look like when the tilt functionality is gone:

<img width="669" height="619" alt="image" src="https://github.com/user-attachments/assets/5d94f794-2111-4ecd-b83e-576d3f9fe242" />

uvNikita commented:
Once you get to the point where you don't have tilt controls on your entities, and if it still shows incorrect percentage, post the diagnostics information for you device. Maybe you have a different firmware version where they changed something.

Orange-GT3 commented:
After a reboot of HA, the tilt entities have gone, as you predicted @uvNikita, thank you.

I'll time the in and out movements next and set those. Just so I am clear, I am putting those times into the two settings circled below?

<img width="425" height="431" alt="image" src="https://github.com/user-attachments/assets/fa8662f3-78b4-432c-9ef8-9c41bad3d58f" />

uvNikita commented:
> I'll time the in and out movements next and set those. Just so I am clear, I am putting those times into the two settings circled below?

Yes, that's correct.
